Apple Silicon を搭載した MacBook で起こる Docker 関連エラーの対処法を追記 (#73)
nnn-training/faq: MARKDOWN/intro/docker.md

これによって Virtualization Framework を無効化でき、エラーが解消されるはずです。

なお、以前より Docker を利用した際の挙動が重くなる可能性があります。

## (6) `yarn add prisma@4.11.0` で失敗する<a id="6"></a>

### 問題詳細
Apple Silicon を搭載した M シリーズ MacBook において、 Docker コンテナ内で `yarn add prisma@4.11.0` を実行すると、以下のようなエラーが表示される場合があります。

```sh
qemu: uncaught target signal 11 (Segmentation fault) - core dumped
Segmentation fault
```

この場合、 Docker Desktop の設定を変更することで解決が望めます。

### →解決方法

Rosetta という機能を有効にするために、Docker Desktop の設定を変更します。

> 以下のスクショで示された右上赤枠内の歯車アイコンをクリックしてください。

Reviewer: ```suggestion
以下のスクリーンショットで示された右上赤枠内の歯車アイコンをクリックしてください。
```

![](./images/docker-enable-rosetta-01.png)

> 次に、 [Use Rosetta for x86/amd64 emulation on Apple Silicon] と書かれた場所があると思います。<br>
> そのチェックボックスをクリックして有効化してください。

Reviewer: スクリーンショットに当該の場所はあるので言い切ってしまった方が読みやすいです。
また `[` と `]` は余分そうです。

```suggestion
次に、Use Rosetta for x86/amd64 emulation on Apple Silicon と書かれたチェックボックスにチェックを入れ、Rosetta を有効化してください。
```

Author: 確かにスクリーンショット内で示されているならば、言い切ってしまって問題ありませんね。
ご指摘ありがとうございます！

![](./images/docker-enable-rosetta-02.png)

> チェックボックスにチェックが入りましたら、右下の [Apply & Restart] をクリックして、 Docker Desktop を再起動してください。

Reviewer: ```suggestion
チェックボックスにチェックを入れたら、右下の「Apply & Restart」と書かれたボタンをクリックして、 Docker Desktop を再起動してください。
```

Author: ボタンと明示するべきでした。
ご指摘ありがとうございます！

![](./images/docker-enable-rosetta-03.png)

これで Rosetta が有効になり、エラーが解消されるはずです。

再起動後、 Docker コンテナを `docker compose up -d --build` コマンドで立ち上げてエラーが解消されたか確認してみてください。

### 原因

> これは、 Docker イメージのアーキテクチャの違いによって起きるエラーです。

Reviewer: 「アーキテクチャ」の語を Docker イメージに対して適用するのは（意味は分かりますが）やや違和感があります。
おそらく [`--platform` のドキュメント](https://docs.docker.com/build/building/multi-platform/#building-multi-platform-images) の `target platform` のことだと思うので、

```suggestion
これは、 Docker イメージが想定するアーキテクチャと実際のコンピュータのアーキテクチャの違いによって起きるエラーです。
```

とするのが自然な気がします。

Author: 意図しているのは `target platform` ですので、ご指摘の通りです。

>これは、 Docker イメージのアーキテクチャの違いによって起きるエラーです。

では、解釈に幅があってよろしくない表現ですね。

提示していただいた記述であれば platform を指定したニュアンスが含まれていて明瞭に感じます。

> 教材では、 `Dockerfile` の 1 行目に `--platform=linux/x86_64` と書かれているかと思います。

Reviewer: ```suggestion
教材では、 `Dockerfile` の 1 行目に `--platform=linux/x86_64` と書かれています。
```

これが Docker イメージの OS とアーキテクチャを指定するための記述です。

[Docker Hub の Node イメージのタグ一覧](https://hub.docker.com/_/node/tags)を見ると、`OS/ARCH` という部分があるかと思います。<br>
実は Docker イメージは OS とアーキテクチャの組み合わせによって分けられており、この部分がそれを示しています。

![](./images/docker-os-arch.png)

`x86_64` というのは Intel CPU 用のアーキテクチャで、 `AMD64` とも言われるものです。一方で、M シリーズ Mac の CPU は ARM アーキテクチャを採用しています。<br>

> つまり、 ARM アーキテクチャで動く CPU 上で AMD64 アーキテクチャを動かさなければなりません。

Reviewer: 一般にはそうではない（ARM で動かしていれば ARM 向けイメージが動く）ので：

```suggestion
つまり、`--platform=linux/x86_64` を指定すると、ARM アーキテクチャで動く CPU 上で AMD64 アーキテクチャを動かさなければなりません。
```

cf. https://docs.docker.jp/engine/reference/builder.html#id39

Author: 前提となる条件が抜けていました。
特定の OS/ARCH を指定した時のみ、異なるアーキテクチャで動かさねばならないことを含めるべきでした。
ご指摘ありがとうございます！

> このための仕組みとしてエミュレータが用意されており、変換して動いてくれるようになっています。

Reviewer: 何を変換してくれているのかを書いた方がいいかもしれません。とはいえ機械語の話をしてもここでは過剰なので：

```suggestion
このための仕組みとしてエミュレータが用意されており、Docker イメージの対応するアーキテクチャと実際のコンピュータのアーキテクチャの差異を吸収してくれるようになっています。
```

Author: 変換だけでは曖昧で疑問を持たせてしまいますね……。
提示していただいた記述であればエミュレータの存在意義が明確に感じられます！
ご指摘ありがとうございます！

そして今回のエラーは、使っているエミュレータがうまく動作していないことが原因です。

ですので、エミュレータを Rosetta というものに切り替えることで解消します。

## (7) M シリーズ Mac にて Docker コンテナ立ち上げ時に警告が出る<a id="7"></a>

### 問題詳細

M シリーズ Mac にて Docker コンテナ立ち上げ時に以下のような警告が出る場合があります。

```sh
app The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
```

### →解決方法

これは致命的なエラーではないので、教材の進行に問題はありませんので無視しても大丈夫です。

> 教材では、 `Dockerfile` の 1 行目に `--platform=linux/x86_64` と書かれた記述を削除し、 `docker compose up -d --build` コマンドを実行すれば解決するはずです。

Reviewer: ```suggestion
教材で提示された `Dockerfile` の 1 行目にある `--platform=linux/x86_64` を削除し、 `docker compose up -d --build` コマンドを実行すれば解決するはずです。
```

### 原因

これは [(6) `yarn add prisma@4.11.0` で失敗する](#6)と同じことが原因です。

> 使っているパソコンのアーキテクチャと呼ばれるものが教材で指定したアーキテクチャと異なるため、パフォーマンスが低下する可能性があり、それを警告として表示しているものです。

Reviewer: ```suggestion
使っているパソコンのアーキテクチャが Docker イメージが想定するものと異なるため、パフォーマンスが低下する可能性があり、それを警告として表示しているものです。
```

Author: Docker イメージが想定するものとした方が冗長さが排除できて良いですね。
ありがとうございます！
